### ISSUE TITLE
Column pruning not handled properly

### ISSUE THREAD

**Issue by birdstorm** [pingcap/tispark#182]
<a href="https://github.com/Novemser"><img src="https://avatars1.githubusercontent.com/u/12718118?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [Novemser](https://github.com/Novemser)**
_Tue Nov 28 07:51:53 2017_
_Originally opened as https://github.com/pingcap/tikv-client-lib-java/issues/163_

----

SQL:
```sql
select id_dt from full_data_type_table where id_dt not in (select id_dt from full_data_type_table  where tp_decimal <> 1E10)
```

Plan:
```bash
== Physical Plan ==
*Project [id_dt#138L]
+- BroadcastNestedLoopJoin BuildRight, LeftAnti, ((id_dt#138L = id_dt#138L#258L) || isnull((id_dt#138L = id_dt#138L#258L)))
   :- TiDB CoprocessorRDD{[table: full_data_type_table] , Ranges: Start:[-9223372036854775808], End: [9223372036854775807], Columns: [id_dt], [tp_varchar], [tp_datetime], [tp_blob], [tp_binary], [tp_date], [tp_timestamp], [tp_year], [tp_bigint], [tp_decimal], [tp_double], [tp_float], [tp_int], [tp_mediumint], [tp_real], [tp_smallint], [tp_tinyint], [tp_char], [tp_nvarchar], [tp_longtext], [tp_mediumtext], [tp_text], [tp_tinytext], [tp_bit], [tp_time], [tp_enum], [tp_set]}
   +- BroadcastExchange IdentityBroadcastMode
      +- *Project [id_dt#138L AS id_dt#138L#258L]
         +- TiDB CoprocessorRDD{[table: full_data_type_table] , Ranges: Start:[-9223372036854775808], End: [9223372036854775807], Columns: [id_dt], [tp_decimal], Aggregates: Not(IsNull([tp_decimal])), <>([tp_decimal], 10000000000)}
```

Seem we didn't properly prune the plan and pushed down too many things.

**Comment by birdstorm** [comment]
<a href="https://github.com/Novemser"><img src="https://avatars1.githubusercontent.com/u/12718118?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Novemser](https://github.com/Novemser)**
_Tue Nov 28 08:05:08 2017_

----

@birdstorm After fixing this issue, regression testing is needed.

**Comment by birdstorm** [comment]
<a href="https://github.com/Novemser"><img src="https://avatars1.githubusercontent.com/u/12718118?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Novemser](https://github.com/Novemser)**
_Fri Dec 15 12:00:42 2017_

----

Schema
```sql
mysql> desc full_data_type_table;
+---------------+-----------------------+------+------+-------------------+-------+
| Field         | Type                  | Null | Key  | Default           | Extra |
+---------------+-----------------------+------+------+-------------------+-------+
| id_dt         | int(11)               | NO   | PRI  | NULL              |       |
| tp_varchar    | varchar(45)           | YES  |      | NULL              |       |
| tp_datetime   | datetime              | YES  |      | CURRENT_TIMESTAMP |       |
| tp_blob       | blob                  | YES  |      | NULL              |       |
| tp_binary     | binary(2)             | YES  |      | NULL              |       |
| tp_date       | date                  | YES  |      | NULL              |       |
| tp_timestamp  | timestamp             | NO   |      | CURRENT_TIMESTAMP |       |
| tp_year       | year UNSIGNED         | YES  |      | NULL              |       |
| tp_bigint     | bigint(20)            | YES  |      | NULL              |       |
| tp_decimal    | decimal               | YES  |      | NULL              |       |
| tp_double     | double                | YES  |      | NULL              |       |
| tp_float      | float                 | YES  |      | NULL              |       |
| tp_int        | int(11)               | YES  | MUL  | NULL              |       |
| tp_mediumint  | mediumint(9)          | YES  |      | NULL              |       |
| tp_real       | double                | YES  |      | NULL              |       |
| tp_smallint   | smallint(6)           | YES  |      | NULL              |       |
| tp_tinyint    | tinyint(4)            | YES  |      | NULL              |       |
| tp_char       | char(10)              | YES  |      | NULL              |       |
| tp_nvarchar   | varchar(40)           | YES  |      | NULL              |       |
| tp_longtext   | longtext              | YES  |      | NULL              |       |
| tp_mediumtext | mediumtext            | YES  |      | NULL              |       |
| tp_text       | text                  | YES  |      | NULL              |       |
| tp_tinytext   | tinytext              | YES  |      | NULL              |       |
| tp_bit        | bit(1)                | YES  |      | NULL              |       |
| tp_time       | time                  | YES  |      | NULL              |       |
| tp_enum       | enum('1','2','3','4') | YES  |      | NULL              |       |
| tp_set        | set('a','b','c','d')  | YES  |      | NULL              |       |
+---------------+-----------------------+------+------+-------------------+-------+
27 rows in set (0.01 sec)
```

**Comment by ilovesoup** [comment]
This is not a problem anymore with @Novemser 's fix on TiDAGRequest's column pruning logic right?
Please confirm. @Novemser

**Comment by ilovesoup** [comment]
This one is still a problem but seems a spark problem.
I've tested on Parquet as well and it shows the same plan result without proper column pruning.
The root cause seems, IN subquery turned into broadcast join operator and split projection from table scan. In that case, column pruning info is not available for scan operator.

**Comment by ilovesoup** [comment]
We need to confirm with spark master branch and post a fix if necessary.

**Comment by birdstorm** [comment]
This issue is fixed in Spark 2.3. Will close it when branch support_spark_2.3 is merged.
